// @radr/shadowwire/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofGenerationError = exports.WASMNotSupportedError = exports.NetworkError = exports.TransferError = exports.ProofUploadError = exports.RecipientNotFoundError = exports.InvalidAmountError = exports.InvalidAddressError = exports.InsufficientBalanceError = exports.ShadowWireError = void 0;
class ShadowWireError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ShadowWireError';
        Object.setPrototypeOf(this, ShadowWireError.prototype);
    }
}
exports.ShadowWireError = ShadowWireError;
class InsufficientBalanceError extends ShadowWireError {
    constructor(message = 'Insufficient balance for this operation') {
        super(message);
        this.name = 'InsufficientBalanceError';
        Object.setPrototypeOf(this, InsufficientBalanceError.prototype);
    }
}
exports.InsufficientBalanceError = InsufficientBalanceError;
class InvalidAddressError extends ShadowWireError {
    constructor(address) {
        super(`Invalid Solana address: ${address}`);
        this.name = 'InvalidAddressError';
        Object.setPrototypeOf(this, InvalidAddressError.prototype);
    }
}
exports.InvalidAddressError = InvalidAddressError;
class InvalidAmountError extends ShadowWireError {
    constructor(message = 'Invalid amount specified') {
        super(message);
        this.name = 'InvalidAmountError';
        Object.setPrototypeOf(this, InvalidAmountError.prototype);
    }
}
exports.InvalidAmountError = InvalidAmountError;
class RecipientNotFoundError extends ShadowWireError {
    constructor(recipient) {
        super(`Recipient not found in ShadowPay system: ${recipient}. Try using an external transfer instead.`);
        this.name = 'RecipientNotFoundError';
        Object.setPrototypeOf(this, RecipientNotFoundError.prototype);
    }
}
exports.RecipientNotFoundError = RecipientNotFoundError;
class ProofUploadError extends ShadowWireError {
    constructor(message = 'Failed to upload zero-knowledge proof') {
        super(message);
        this.name = 'ProofUploadError';
        Object.setPrototypeOf(this, ProofUploadError.prototype);
    }
}
exports.ProofUploadError = ProofUploadError;
class TransferError extends ShadowWireError {
    constructor(message = 'Transfer execution failed') {
        super(message);
        this.name = 'TransferError';
        Object.setPrototypeOf(this, TransferError.prototype);
    }
}
exports.TransferError = TransferError;
class NetworkError extends ShadowWireError {
    constructor(message = 'Network request failed') {
        super(message);
        this.name = 'NetworkError';
        Object.setPrototypeOf(this, NetworkError.prototype);
    }
}
exports.NetworkError = NetworkError;
class WASMNotSupportedError extends ShadowWireError {
    constructor() {
        super('WebAssembly is not supported in this environment. Use backend proof generation instead.');
        this.name = 'WASMNotSupportedError';
        Object.setPrototypeOf(this, WASMNotSupportedError.prototype);
    }
}
exports.WASMNotSupportedError = WASMNotSupportedError;
class ProofGenerationError extends ShadowWireError {
    constructor(message = 'Failed to generate zero-knowledge proof') {
        super(message);
        this.name = 'ProofGenerationError';
        Object.setPrototypeOf(this, ProofGenerationError.prototype);
    }
}
exports.ProofGenerationError = ProofGenerationError;
