// @radr/shadowwire/dist/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenUtils = void 0;
const constants_1 = require("./constants");
class TokenUtils {
    static toSmallestUnit(amount, token) {
        const decimals = this.getTokenDecimals(token);
        return Math.floor(amount * Math.pow(10, decimals));
    }
    static fromSmallestUnit(amount, token) {
        const decimals = this.getTokenDecimals(token);
        return amount / Math.pow(10, decimals);
    }
    static getTokenDecimals(token) {
        const decimals = constants_1.TOKEN_DECIMALS[token];
        if (decimals === undefined) {
            throw new Error(`Unknown token: ${token}`);
        }
        return decimals;
    }
    static getTokenMint(token) {
        const mint = constants_1.TOKEN_MINTS[token];
        if (!mint) {
            throw new Error(`Unknown token: ${token}`);
        }
        return mint;
    }
    static isValidToken(token) {
        return token in constants_1.TOKEN_DECIMALS;
    }
    static getAllTokens() {
        return Object.keys(constants_1.TOKEN_DECIMALS);
    }
}
exports.TokenUtils = TokenUtils;
