// @radr/shadowwire/dist/zkProofs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.initWASM = initWASM;
exports.generateRangeProof = generateRangeProof;
exports.verifyRangeProof = verifyRangeProof;
exports.isWASMSupported = isWASMSupported;
const settler_wasm_1 = __importStar(require("../wasm/settler_wasm"));
const errors_1 = require("./errors");
let wasmInitialized = false;
/**
 * Detects if running in Node.js environment
 */
function isNode() {
    return typeof process !== 'undefined' &&
        process.versions != null &&
        process.versions.node != null;
}
/**
 * Initializes WASM module for Node.js environment
 */
async function initWASMNode() {
    // Dynamic import for Node.js-only modules
    const path = await Promise.resolve().then(() => __importStar(require('path')));
    const fs = await Promise.resolve().then(() => __importStar(require('fs')));
    const wasmPaths = [
        path.join(__dirname, '../wasm/settler_wasm_bg.wasm'),
        path.join(process.cwd(), 'wasm/settler_wasm_bg.wasm'),
        path.join(process.cwd(), 'dist/wasm/settler_wasm_bg.wasm'),
        path.join(process.cwd(), 'node_modules/@radr/shadowwire/dist/wasm/settler_wasm_bg.wasm'),
    ];
    let wasmPath;
    for (const p of wasmPaths) {
        if (fs.existsSync(p)) {
            wasmPath = p;
            break;
        }
    }
    if (!wasmPath) {
        throw new Error('WASM file not found in any expected location');
    }
    const wasmBuffer = fs.readFileSync(wasmPath);
    await (0, settler_wasm_1.default)(wasmBuffer);
}
/**
 * Initializes WASM module for browser environment
 */
async function initWASMBrowser(wasmUrl) {
    // Default WASM URLs to try in browser
    const defaultUrls = [
        '/wasm/settler_wasm_bg.wasm',
        './wasm/settler_wasm_bg.wasm',
        '../wasm/settler_wasm_bg.wasm',
    ];
    const urlsToTry = wasmUrl ? [wasmUrl, ...defaultUrls] : defaultUrls;
    let lastError = null;
    for (const url of urlsToTry) {
        try {
            const response = await fetch(url);
            if (!response.ok) {
                throw new Error(`Failed to fetch WASM: ${response.status} ${response.statusText}`);
            }
            const wasmBuffer = await response.arrayBuffer();
            await (0, settler_wasm_1.default)(wasmBuffer);
            return; // Success!
        }
        catch (error) {
            lastError = error;
            // Continue to next URL
        }
    }
    throw new Error(`Could not load WASM from any location. Last error: ${lastError?.message}. ` +
        `In browser environments, ensure the WASM file is served at one of: ${urlsToTry.join(', ')}`);
}
/**
 * Initializes the WASM module. Must be called before using proof functions.
 *
 * @param wasmUrl - (Browser only) Optional custom URL to the WASM file.
 *                  If not provided, will try default paths: /wasm/settler_wasm_bg.wasm,
 *                  ./wasm/settler_wasm_bg.wasm, ../wasm/settler_wasm_bg.wasm
 *
 * @example
 * // Node.js
 * await initWASM();
 *
 * @example
 * // Browser with default path
 * await initWASM();
 *
 * @example
 * // Browser with custom path
 * await initWASM('/public/wasm/settler_wasm_bg.wasm');
 */
async function initWASM(wasmUrl) {
    if (wasmInitialized) {
        return;
    }
    try {
        if (isNode()) {
            await initWASMNode();
        }
        else {
            await initWASMBrowser(wasmUrl);
        }
        wasmInitialized = true;
    }
    catch (error) {
        throw new errors_1.ProofGenerationError(`Could not load cryptography module: ${error.message}`);
    }
}
async function generateRangeProof(amount, bitLength = 64) {
    if (!wasmInitialized) {
        await initWASM();
    }
    if (amount < 0) {
        throw new errors_1.ProofGenerationError('Amount must be non-negative');
    }
    const maxAmount = Math.pow(2, bitLength);
    if (amount >= maxAmount) {
        throw new errors_1.ProofGenerationError(`Amount exceeds ${bitLength}-bit range`);
    }
    if (!Number.isInteger(amount)) {
        throw new errors_1.ProofGenerationError('Amount must be an integer');
    }
    try {
        const result = (0, settler_wasm_1.generate_range_proof)(BigInt(amount), bitLength);
        return {
            proofBytes: uint8ArrayToHex(result.proof_bytes),
            commitmentBytes: uint8ArrayToHex(result.commitment_bytes),
            blindingFactorBytes: uint8ArrayToHex(result.blinding_factor_bytes),
        };
    }
    catch (error) {
        throw new errors_1.ProofGenerationError(`Failed to generate proof: ${error.message || error}`);
    }
}
async function verifyRangeProof(proofBytes, commitmentBytes, bitLength = 64) {
    if (!wasmInitialized) {
        await initWASM();
    }
    try {
        const proofArray = hexToUint8Array(proofBytes);
        const commitmentArray = hexToUint8Array(commitmentBytes);
        return (0, settler_wasm_1.verify_range_proof)(proofArray, commitmentArray, bitLength);
    }
    catch (error) {
        return false;
    }
}
function isWASMSupported() {
    try {
        return typeof WebAssembly === 'object' &&
            typeof WebAssembly.instantiate === 'function';
    }
    catch (e) {
        return false;
    }
}
function uint8ArrayToHex(bytes) {
    return Array.from(bytes)
        .map(byte => byte.toString(16).padStart(2, '0'))
        .join('');
}
function hexToUint8Array(hex) {
    if (hex.length % 2 !== 0) {
        throw new Error('Hex string must have even length');
    }
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < hex.length; i += 2) {
        bytes[i / 2] = parseInt(hex.substr(i, 2), 16);
    }
    return bytes;
}
